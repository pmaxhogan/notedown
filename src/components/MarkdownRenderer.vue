<script setup>
import * as DOMPurify from "dompurify";
import { parse } from "marked";
import { ref, watchEffect } from "vue";

const sanitize = DOMPurify.default.sanitize;

const props = defineProps({
  content: {
    type: String,
    required: true,
  },
});

const MAX_CONTENT_LENGTH = 10 * 1000 * 1000;
if (props.content.length > MAX_CONTENT_LENGTH) {
  throw new Error("Content too long");
}

const output = ref("");

watchEffect(() => {
  output.value = sanitize(parse(props.content));
  window.requestAnimationFrame(() => {
    if (typeof window.MathJax !== "undefined") {
      // FIXME: testing does not work
      window.MathJax.typeset();
    }
  });
});
</script>

<template>
  <div v-html="output"></div>
</template>
