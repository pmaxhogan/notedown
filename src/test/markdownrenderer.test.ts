import { mount } from "@vue/test-utils";
import MarkdownRenderer from "../components/MarkdownRenderer.vue";
import { expect, test } from "vitest";

const validString = `
# Hello Markdown!
$$\\frac\\alpha\\beta$$
`.trim();

const invalidString = `
$$\\notanescape$$

\`unclosed code block
`.trim();

const errorString = "a".repeat(10 * 1000 * 1000 + 1);

test("markdown renderer component valid", async () => {
  expect(MarkdownRenderer).toBeTruthy();

  const wrapper = mount(MarkdownRenderer, {
    props: {
      content: validString,
    },
  });

  expect(wrapper.find("#hello-markdown").text()).toBe("Hello Markdown!");
  expect(wrapper.html()).toMatchSnapshot();
});

test("markdown renderer component invalid", async () => {
  expect(MarkdownRenderer).toBeTruthy();

  const wrapper = mount(MarkdownRenderer, {
    props: {
      content: invalidString,
    },
  });

  expect(wrapper.find("code").exists()).toBe(false);
  expect(wrapper.html()).toMatchSnapshot();
});

test("markdown renderer component error", async () => {
  expect(MarkdownRenderer).toBeTruthy();

  expect(() =>
    mount(MarkdownRenderer, {
      props: {
        content: errorString,
      },
    })
  ).toThrow();
});
