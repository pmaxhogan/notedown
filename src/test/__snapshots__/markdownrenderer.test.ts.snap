// Vitest Snapshot v1

exports[`markdown renderer component invalid 1`] = `
"<div>
  <p>$$\\\\notanescape$$</p>
  <p>\`unclosed code block</p>
</div>"
`;

exports[`markdown renderer component valid 1`] = `
"<div>
  <h1 id=\\"hello-markdown\\">Hello Markdown!</h1>
  <p>$$\\\\frac\\\\alpha\\\\beta$$</p>
</div>"
`;
